test-workloads/src/bin/tellus.rs

```diff
@@ -300,6 +411,34 @@ extern "C" fn kernel_init(hart_id: u64, fdt_addr: u64) {
         mem_range.base(),
         mem_range.size()
     );
+    let mut next_page = (mem_range.base() + mem_range.size() / 2) & !0x3fff;
+
+    // Initialize CPUs.
+    PER_CPU.call_once(|| {
```

Comment:
Hm? You set PER_CPU only when this block is executed and returns CPUs. But then you start the CPUs here, creating a race condition as the first thing the CPU does is calling PER_CPU.get().

I think this is shown in the test failures, perhaps (haven't checked) because QEMU is been running with the multi threaded accel, or we're just lucky on this instance.

Reply:
Ah, good catch. No idea why I never ended up triggering this locally... will fix.